### PR TITLE
Vault package

### PR DESCRIPTION
**Pull request by shlomitk1**
Moved all vault related functionality to pkg/vault.

### CODE REVIEW

**Comment by revit13** [comment]
Relates to issue https://github.com/IBM/the-mesh-for-data/issues/319

File: pkg/vault/vault_interface.go [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+// GetVaultAuthTTL returns the amount of time the authorization issued by vault is valid
+func GetVaultAuthTTL() string {
+	return os.Getenv("VAULT_TTL")
+}
+
+// GetVaultAuthService returns the authentication service that was chosen for use with vault,
+// and the configuration options for it.
+// Vault support multiple different types of authentication such as java web tokens (jwt), github, aws ...
+func GetVaultAuthService() (string, api.EnableAuthOptions) {
+	auth := os.Getenv("VAULT_AUTH")
```
Comment (revit13) [inline pkg/vault/vault_interface.go:279]: Consider getting auth path as a parameter to this function.

File: pkg/vault/vault_interface.go [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+	if err != nil {
+		msg := "Error creating request to mount vault for user credentials " + url + ":" + err.Error()
+		return errors.New(msg)
+	}
+	req.Header.Set("X-Vault-Token", c.Token)
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		msg := "Error mounting vault for user credentials " + url + ":" + err.Error()
```
Comment (revit13) [inline pkg/vault/vault_interface.go:225]: same here.. user creds should not be mentioned in the error...

File: pkg/vault/vault_interface.go [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+	}
+}
+
+// NewDummyConnection returns a new Dummy object
+func NewDummyConnection() *Dummy {
+	return &Dummy{}
+}
+
+// CredentialManagerInterface provides vault functionality
+type CredentialManagerInterface interface {
```
Comment (revit13) [inline pkg/vault/vault_interface.go:47]: consider moving this interface to another file as its not vault specific. or rename the package name to something like secretmanager as it contains the dummy implementation as well.

**Review by revit13 — APPROVED** [review]

File: pkg/vault/vault_interface.go [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+	logicalClient := c.Client.Logical()
+	if logicalClient == nil {
+		msg := "No logical client received when linking policy " + policyName + " to idenity " + identity
+		return errors.New(msg)
+	}
+
+	params := map[string]interface{}{
+		"user_claim":                       "sub",
+		"role_type":                        auth,
+		"bound_service_account_names":      "secret-provider",
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply pkg/vault/vault_interface.go]: Done

File: pkg/vault/vault_interface.go [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+}
+
+// Mount mounts a key-value secret provider (kv version 1) to manage the storage of the secrets
+func (c *Connection) Mount(path string) error {
+	body := strings.NewReader(`{"type":"kv-v1"}`)
+	url := c.Address + path
+
+	req, err := http.NewRequest("POST", url, body)
+	if err != nil {
+		msg := "Error creating request to mount vault for user credentials " + url + ":" + err.Error()
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply pkg/vault/vault_interface.go]: Fixed

File: pkg/vault/vault_interface.go [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+	GetSecret(vaultPath string) (string, error)
+	AddSecret(path string, credentials map[string]interface{}) error
+	AddSecretFromStruct(path string, creds interface{}) error
+}
+
+// LinkVaultPolicyToIdentity registers a policy for a given identity or role, meaning that when a person or service
+// of that identity logs into vault and tries to read or write a secret the provided policy
+// will determine whether that is allowed or not.
+func (c *Connection) LinkVaultPolicyToIdentity(identity string, policyName string, boundedNamespace string) error {
+	auth, _ := GetVaultAuthService()
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply pkg/vault/vault_interface.go]: Done

File: pkg/vault/vault_interface.go [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+	identityPath := GetIdentitiesVaultAuthPath() + "/" + identity
+
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply pkg/vault/vault_interface.go]: Done

File: pkg/vault/vault_interface.go [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+		return errors.New(msg)
+	}
+
+	return nil
+}
+
+// RemoveVaultPolicyFromIdentity removes the policy from the authentication identity with which it is associated, meaning
+// this policy will no longer be invoked when a person or service authenticates with this identity.
+func (c *Connection) RemoveVaultPolicyFromIdentity(identity string, policyName string) error {
+	identityPath := GetIdentitiesVaultAuthPath() + "/" + identity
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply pkg/vault/vault_interface.go]: Done

File: pkg/vault/vault_interface.go [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+		return errors.New(msg)
+	}
+
+	params := map[string]interface{}{
+		"user_claim":                       "sub",
+		"role_type":                        auth,
+		"bound_service_account_names":      "secret-provider",
+		"bound_service_account_namespaces": boundedNamespace,
+		"policies":                         policyName,
+		"ttl":                              GetVaultAuthTTL(),
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply pkg/vault/vault_interface.go]: done

File: pkg/vault/vault_interface.go [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+	}
+	defer resp.Body.Close()
+
+	return nil
+}
+
+// GetIdentitiesVaultAuthPath returns the path where policies for the different identities are written
+// Includes the namespace so that multiple m4d control planes can use the same vault instance.
+func GetIdentitiesVaultAuthPath() string {
+	auth, _ := GetVaultAuthService()
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply pkg/vault/vault_interface.go]: done

File: pkg/vault/vault_interface.go (hunk 1 above)
Comment (shlomitk1) [reply pkg/vault/vault_interface.go]: done

File: pkg/vault/vault_interface.go (hunk 2 above)
Comment (shlomitk1) [reply pkg/vault/vault_interface.go]: fixed

File: pkg/vault/vault_interface.go (hunk 3 above)
Comment (shlomitk1) [reply pkg/vault/vault_interface.go]: Split into several files, renamed the interface name

**Pull request merged**